File: frontend/src/services/RemoteDesktopManager.ts
/**
 * Remote Desktop Manager
 * 
 * Handles screen sharing and remote control functionality
 */
export default class RemoteDesktopManager {
  private peerConnection: RTCPeerConnection | null = null;
  private localStream: MediaStream | null = null;
  private remoteStream: MediaStream | null = null;
  private dataChannel: RTCDataChannel | null = null;
  private ws: WebSocket;
  private sessionId: string;
  private sourceDeviceId: string;
  private viewerDeviceId: string;
  private isSource: boolean;
  private onRemoteStream: ((stream: MediaStream) => void) | null = null;
  private onConnectionStateChange: ((state: string) => void) | null = null;

  constructor(
    ws: WebSocket,
    sessionId: string,
    sourceDeviceId: string,
    viewerDeviceId: string,
    isSource: boolean
  ) {
    this.ws = ws;
    this.sessionId = sessionId;
    this.sourceDeviceId = sourceDeviceId;
    this.viewerDeviceId = viewerDeviceId;
    this.isSource = isSource;

    // Listen for WebRTC signaling
    const messageHandler = (event: MessageEvent) => {
      try {
        const message = JSON.parse(event.data);
        // Handle WebRTC signaling messages
        // For source: messages come from viewerDeviceId
        // For viewer: messages come from sourceDeviceId
        const expectedFromDevice = this.isSource ? this.viewerDeviceId : this.sourceDeviceId;
        
        if (message.type.startsWith('webrtc_')) {
          const fromDevice = message.deviceId || message.payload?.fromDevice;
          if (fromDevice === expectedFromDevice) {
            this.handleSignaling(message);
          }
        }
      } catch (error) {
        // Ignore non-JSON messages or parsing errors
      }
    };
    
    ws.addEventListener('message', messageHandler);
    
    // Store handler for cleanup
    (this as any)._messageHandler = messageHandler;
  }

  /**
   * Start screen sharing (call on source device)
   */
  async startScreenShare(): Promise<MediaStream> {
    try {
      console.log('Requesting screen share permission...');
      
      // Request screen capture - this will show browser permission dialog
      // The user must interact with the page for this to work (user gesture required)
      this.localStream = await navigator.mediaDevices.getDisplayMedia({
        video: {
          cursor: 'always',
          displaySurface: 'monitor',
        } as MediaTrackConstraints,
        audio: false,
      });

      console.log('Screen share permission granted, stream received');

      // Handle stream end (user stops sharing)
      this.localStream.getVideoTracks()[0].addEventListener('ended', () => {
        console.log('Screen share ended by user');
        this.stopScreenShare();
      });

      // Create peer connection
      await this.createPeerConnection();

      // Add local stream tracks
      this.localStream.getTracks().forEach(track => {
        this.peerConnection!.addTrack(track, this.localStream!);
      });

      // Create offer
      const offer = await this.peerConnection!.createOffer({
        offerToReceiveAudio: false,
        offerToReceiveVideo: false,
      });
      await this.peerConnection!.setLocalDescription(offer);

      console.log('Sending WebRTC offer for remote desktop');

      // Send offer via WebSocket
      if (this.ws.readyState === WebSocket.OPEN) {
        this.ws.send(JSON.stringify({
          type: 'webrtc_offer',
          sessionId: this.sessionId,
          deviceId: this.sourceDeviceId,
          payload: {
            toDevice: this.viewerDeviceId,
            data: offer,
            purpose: 'remote_desktop',
          },
          timestamp: Date.now(),
        }));
      } else {
        throw new Error('WebSocket not open, cannot send offer');
      }

      return this.localStream;
    } catch (error) {
      console.error('Failed to start screen share:', error);
      if (error instanceof Error) {
        if (error.name === 'NotAllowedError' || error.name === 'NotReadableError') {
          throw new Error('Screen share permission denied or not available');
        }
        throw error;
      }
      throw new Error('Unknown error starting screen share');
    }
  }

  /**
   * Connect as viewer (call on viewer device)
   */
  async connectAsViewer(): Promise<void> {
    await this.createPeerConnection();

    // Set up remote stream handler
    this.peerConnection!.ontrack = (event) => {
      if (event.streams && event.streams[0]) {
        this.remoteStream = event.streams[0];
        if (this.onRemoteStream) {
          this.onRemoteStream(this.remoteStream);
        }
      }
    };

    // Create answer when offer is received (handled in handleSignaling)
  }

  /**
   * Create WebRTC peer connection
   */
  private async createPeerConnection(): Promise<void> {
    this.peerConnection = new RTCPeerConnection({
      iceServers: [
        { urls: 'stun:stun.l.google.com:19302' },
        { urls: 'stun:stun1.l.google.com:19302' },
      ],
    });

    // Handle ICE candidates
    this.peerConnection.onicecandidate = (event) => {
      if (event.candidate) {
        this.ws.send(JSON.stringify({
          type: 'webrtc_ice_candidate',
          sessionId: this.sessionId,
          deviceId: this.isSource ? this.sourceDeviceId : this.viewerDeviceId,
          payload: {
            toDevice: this.isSource ? this.viewerDeviceId : this.sourceDeviceId,
            data: event.candidate.toJSON(),
          },
          timestamp: Date.now(),
        }));
      }
    };

    // Handle connection state changes
    this.peerConnection.onconnectionstatechange = () => {
      const state = this.peerConnection?.connectionState || 'disconnected';
      if (this.onConnectionStateChange) {
        this.onConnectionStateChange(state);
      }
    };

    // Create data channel for input events
    if (this.isSource) {
      // Source device creates data channel
      this.dataChannel = this.peerConnection.createDataChannel('input', {
        ordered: true,
      });
      this.setupDataChannel();
    } else {
      // Viewer device receives data channel
      this.peerConnection.ondatachannel = (event) => {
        this.dataChannel = event.channel;
        this.setupDataChannel();
      };
    }
  }

  /**
   * Setup data channel for input events
   */
  private setupDataChannel(): void {
    if (!this.dataChannel) return;

    this.dataChannel.onopen = () => {
      console.log('Data channel opened for remote desktop');
    };

    this.dataChannel.onerror = (error) => {
      console.error('Data channel error:', error);
    };

    if (this.isSource) {
      // Source device receives input events from viewer
      this.dataChannel.onmessage = (event) => {
        try {
          const message = JSON.parse(event.data);
          if (message.type === 'input_event') {
            this.handleInputEvent(message.event);
          }
        } catch (error) {
          console.error('Failed to parse input event:', error);
        }
      };
    } else {
      // Viewer device listens for messages (status updates, etc.)
      this.dataChannel.onmessage = (event) => {
        try {
          const message = JSON.parse(event.data);
          console.log('Data channel message:', message);
        } catch (error) {
          console.error('Failed to parse data channel message:', error);
        }
      };
    }
  }

  /**
   * Handle WebRTC signaling messages
   */
  private async handleSignaling(message: any): Promise<void> {
    if (!this.peerConnection) {
      console.log('Creating peer connection for signaling');
      await this.createPeerConnection();
    }

    const { data } = message.payload;

    try {
      if (message.type === 'webrtc_offer') {
        // Viewer receives offer from source
        console.log('Received WebRTC offer');
        await this.peerConnection.setRemoteDescription(new RTCSessionDescription(data));
        const answer = await this.peerConnection.createAnswer();
        await this.peerConnection.setLocalDescription(answer);

        this.ws.send(JSON.stringify({
          type: 'webrtc_answer',
          sessionId: this.sessionId,
          deviceId: this.viewerDeviceId,
          payload: {
            toDevice: this.sourceDeviceId,
            data: answer,
          },
          timestamp: Date.now(),
        }));
        console.log('Sent WebRTC answer');
      } else if (message.type === 'webrtc_answer') {
        // Source receives answer from viewer
        console.log('Received WebRTC answer');
        await this.peerConnection.setRemoteDescription(new RTCSessionDescription(data));
      } else if (message.type === 'webrtc_ice_candidate') {
        // Handle ICE candidate
        if (data) {
          await this.peerConnection.addIceCandidate(new RTCIceCandidate(data));
        }
      }
    } catch (error) {
      console.error('Error handling WebRTC signaling:', error);
    }
  }

  /**
   * Send input event to source device (call from viewer)
   */
  sendInputEvent(event: {
    type: 'mousedown' | 'mouseup' | 'mousemove' | 'click' | 'keydown' | 'keyup' | 'scroll';
    x?: number;
    y?: number;
    button?: number;
    key?: string;
    code?: string;
    deltaX?: number;
    deltaY?: number;
  }): void {
    if (this.dataChannel && this.dataChannel.readyState === 'open') {
      this.dataChannel.send(JSON.stringify({
        type: 'input_event',
        event,
        timestamp: Date.now(),
      }));
    }
  }

  /**
   * Set callback for remote stream
   */
  setOnRemoteStream(callback: (stream: MediaStream) => void): void {
    this.onRemoteStream = callback;
  }

  /**
   * Set callback for connection state changes
   */
  setOnConnectionStateChange(callback: (state: string) => void): void {
    this.onConnectionStateChange = callback;
  }

  /**
   * Stop screen sharing
   */
  stopScreenShare(): void {
    if (this.localStream) {
      this.localStream.getTracks().forEach(track => track.stop());
      this.localStream = null;
    }
    if (this.peerConnection) {
      this.peerConnection.close();
      this.peerConnection = null;
    }
    if (this.dataChannel) {
      this.dataChannel.close();
      this.dataChannel = null;
    }
  }

  /**
   * Handle input event on source device
   * 
   * Note: Browser security prevents direct input simulation.
   * For full functionality, use:
   * 1. Electron app with native input APIs
   * 2. Browser extension with native messaging
   * 3. Desktop application wrapper
   */
  private handleInputEvent(event: {
    type: string;
    x?: number;
    y?: number;
    button?: number;
    key?: string;
    code?: string;
    deltaX?: number;
    deltaY?: number;
  }): void {
    console.log('Received input event:', event);
    
    // Dispatch custom event for browser extension/Electron wrapper
    window.dispatchEvent(new CustomEvent('remote_input_event', {
      detail: event,
    }));

    // Try to simulate events using available browser APIs
    // Note: This has severe limitations due to browser security policies
    
    if (event.type === 'mousedown' || event.type === 'mouseup' || event.type === 'click') {
      if (event.x !== undefined && event.y !== undefined) {
        // Create mouse event
        const mouseEvent = new MouseEvent(event.type, {
          view: window,
          bubbles: true,
          cancelable: true,
          clientX: event.x,
          clientY: event.y,
          button: event.button || 0,
        });
        
        // Try to find element at coordinates and dispatch event
        const element = document.elementFromPoint(event.x, event.y);
        if (element) {
          element.dispatchEvent(mouseEvent);
        } else {
          // Fallback: dispatch on document
          document.dispatchEvent(mouseEvent);
        }
      }
    } else if (event.type === 'mousemove') {
      if (event.x !== undefined && event.y !== undefined) {
        const mouseEvent = new MouseEvent('mousemove', {
          view: window,
          bubbles: true,
          cancelable: true,
          clientX: event.x,
          clientY: event.y,
        });
        document.dispatchEvent(mouseEvent);
      }
    } else if (event.type === 'scroll') {
      if (event.deltaX !== undefined || event.deltaY !== undefined) {
        const wheelEvent = new WheelEvent('wheel', {
          view: window,
          bubbles: true,
          cancelable: true,
          clientX: event.x || 0,
          clientY: event.y || 0,
          deltaX: event.deltaX || 0,
          deltaY: event.deltaY || 0,
        });
        document.dispatchEvent(wheelEvent);
      }
    } else if (event.type === 'keydown' || event.type === 'keyup') {
      if (event.key) {
        const keyEvent = new KeyboardEvent(event.type, {
          view: window,
          bubbles: true,
          cancelable: true,
          key: event.key,
          code: event.code || '',
        });
        
        // Try to dispatch on active element first
        if (document.activeElement) {
          document.activeElement.dispatchEvent(keyEvent);
        } else {
          document.dispatchEvent(keyEvent);
        }
      }
    }
  }

  /**
   * Cleanup
   */
  cleanup(): void {
    this.stopScreenShare();
    if (this.remoteStream) {
      this.remoteStream.getTracks().forEach(track => track.stop());
      this.remoteStream = null;
    }
    // Remove message handler
    if ((this as any)._messageHandler) {
      this.ws.removeEventListener('message', (this as any)._messageHandler);
    }
  }
}

File: mobile/android/app/src/main/java/com/flowlink/app/service/RemoteDesktopManager.kt
package com.flowlink.app.service

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.media.projection.MediaProjection
import android.media.projection.MediaProjectionManager
import android.util.Log
import org.json.JSONObject
import org.webrtc.*

/**
 * Remote Desktop Manager for Android
 * 
 * Handles screen sharing and remote control functionality using WebRTC
 */
class RemoteDesktopManager(
    private val context: Context,
    private val webSocketManager: WebSocketManager,
    private val sessionId: String,
    private val sourceDeviceId: String,
    private val viewerDeviceId: String,
    private val isSource: Boolean
) {
    private var peerConnection: PeerConnection? = null
    private var dataChannel: DataChannel? = null
    private var videoCapturer: VideoCapturer? = null
    private var videoSource: VideoSource? = null
    private var videoTrack: VideoTrack? = null
    private var localStream: MediaStream? = null
    // Don't store MediaProjection - create fresh each time to avoid reuse error
    
    private val peerConnectionFactory: PeerConnectionFactory by lazy {
        // Initialize WebRTC
        val initOptions = PeerConnectionFactory.InitializationOptions.builder(context)
            .setEnableInternalTracer(true)
            .setFieldTrials("")
            .createInitializationOptions()
        
        try {
            PeerConnectionFactory.initialize(initOptions)
            Log.d(TAG, "WebRTC initialized successfully")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize WebRTC", e)
        }
        
        val options = PeerConnectionFactory.Options()
        
        val encoderFactory = DefaultVideoEncoderFactory(
            EglBase.create().eglBaseContext,
            true,  // enableIntelVp8Encoder
            true   // enableH264HighProfile
        )
        
        val decoderFactory = DefaultVideoDecoderFactory(EglBase.create().eglBaseContext)
        
        PeerConnectionFactory.builder()
            .setOptions(options)
            .setVideoEncoderFactory(encoderFactory)
            .setVideoDecoderFactory(decoderFactory)
            .createPeerConnectionFactory()
    }
    
    companion object {
        private const val TAG = "RemoteDesktopManager"
        const val SCREEN_CAPTURE_REQUEST_CODE = 1001
    }
    
    /**
     * Request screen capture permission (call from Activity)
     */
    fun requestScreenCapturePermission(activity: Activity) {
        val mediaProjectionManager = context.getSystemService(Context.MEDIA_PROJECTION_SERVICE) as MediaProjectionManager
        val captureIntent = mediaProjectionManager.createScreenCaptureIntent()
        activity.startActivityForResult(captureIntent, SCREEN_CAPTURE_REQUEST_CODE)
    }
    
    /**
     * Start screen sharing after permission granted
     * CRITICAL: This must be called IMMEDIATELY after receiving permission
     * The Intent data can only be used ONCE and cannot be stored/cloned
     */
    fun startScreenShare(resultCode: Int, data: Intent) {
        Log.d(TAG, "=== START SCREEN SHARE ===")
        Log.d(TAG, "Result code: $resultCode")
        
        var mediaProjection: MediaProjection? = null
        
        try {
            // Validate result code first
            if (resultCode != Activity.RESULT_OK) {
                throw Exception("Invalid result code: $resultCode")
            }
            
            // Clean up existing resources
            if (videoCapturer != null || videoTrack != null || peerConnection != null) {
                Log.d(TAG, "Cleaning up existing resources...")
                stopScreenShare()
                Thread.sleep(300)
            }
            
            // Get MediaProjectionManager
            val mediaProjectionManager = context.getSystemService(Context.MEDIA_PROJECTION_SERVICE) as MediaProjectionManager
            
            // CRITICAL: Get MediaProjection IMMEDIATELY
            // Do NOT store, clone, or delay this call
            Log.d(TAG, "Getting MediaProjection IMMEDIATELY...")
            try {
                mediaProjection = mediaProjectionManager.getMediaProjection(resultCode, data)
                if (mediaProjection == null) {
                    throw Exception("MediaProjection returned null")
                }
            } catch (e: IllegalStateException) {
                Log.e(TAG, "IllegalStateException getting MediaProjection", e)
                throw Exception("MediaProjection failed: ${e.message}. The permission may have expired or is already in use.")
            } catch (e: Exception) {
                Log.e(TAG, "Exception getting MediaProjection", e)
                throw Exception("Failed to get MediaProjection: ${e.message}")
            }
            
            Log.d(TAG, "✅ MediaProjection obtained successfully")
            
            // Register callback for when projection stops
            val mediaProjectionCallback = object : MediaProjection.Callback() {
                override fun onStop() {
                    Log.d(TAG, "MediaProjection stopped by system")
                    stopScreenShare()
                }
            }
            mediaProjection.registerCallback(mediaProjectionCallback, null)
            
            // Create peer connection
            Log.d(TAG, "Creating peer connection...")
            createPeerConnection()
            
            // CRITICAL: Create ScreenCapturerAndroid with the ORIGINAL Intent
            // Pass the Intent DIRECTLY - do not clone or modify
            Log.d(TAG, "Creating ScreenCapturerAndroid...")
            videoCapturer = ScreenCapturerAndroid(data, mediaProjectionCallback)
            Log.d(TAG, "✅ ScreenCapturerAndroid created")
            
            // Create video source and track
            Log.d(TAG, "Creating video source...")
            val surfaceTextureHelper = SurfaceTextureHelper.create("CaptureThread", EglBase.create().eglBaseContext)
            videoSource = peerConnectionFactory.createVideoSource(videoCapturer!!.isScreencast)
            
            Log.d(TAG, "Initializing video capturer...")
            videoCapturer!!.initialize(
                surfaceTextureHelper,
                context,
                videoSource!!.capturerObserver
            )
            
            Log.d(TAG, "Creating video track...")
            videoTrack = peerConnectionFactory.createVideoTrack("screen_video_${System.currentTimeMillis()}", videoSource)
            videoTrack!!.setEnabled(true)
            
            // Create local stream and add track
            Log.d(TAG, "Creating local stream...")
            localStream = peerConnectionFactory.createLocalMediaStream("screen_stream_${System.currentTimeMillis()}")
            localStream!!.addTrack(videoTrack)
            
            // Add stream to peer connection
            Log.d(TAG, "Adding stream to peer connection...")
            peerConnection?.addStream(localStream)
            
            // Start capturing
            Log.d(TAG, "Starting video capture at 1280x720@30fps...")
            videoCapturer!!.startCapture(1280, 720, 30)
            
            // Create and send offer
            Log.d(TAG, "Creating and sending WebRTC offer...")
            createAndSendOffer()
            
            Log.d(TAG, "Screen sharing started successfully")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to start screen share", e)
            // Clean up on error
            mediaProjection?.stop()
            stopScreenShare()
            throw e
        }
    }
    
    /**
     * Create screen capturer using MediaProjection
     * Note: This method is no longer used as we create the capturer directly in startScreenShare
     */
    @Deprecated("Use startScreenShare directly")
    private fun createScreenCapturer(): VideoCapturer {
        // This is kept for reference but not used
        throw UnsupportedOperationException("Use startScreenShare directly")
    }
    
    /**
     * Create WebRTC peer connection
     */
    private fun createPeerConnection() {
        val iceServers = listOf(
            PeerConnection.IceServer.builder("stun:stun.l.google.com:19302").createIceServer(),
            PeerConnection.IceServer.builder("stun:stun1.l.google.com:19302").createIceServer()
        )
        
        val rtcConfig = PeerConnection.RTCConfiguration(iceServers).apply {
            bundlePolicy = PeerConnection.BundlePolicy.MAXBUNDLE
            rtcpMuxPolicy = PeerConnection.RtcpMuxPolicy.REQUIRE
            tcpCandidatePolicy = PeerConnection.TcpCandidatePolicy.DISABLED
        }
        
        peerConnection = peerConnectionFactory.createPeerConnection(
            rtcConfig,
            object : PeerConnection.Observer {
                override fun onIceCandidate(candidate: IceCandidate) {
                    sendIceCandidate(candidate)
                }
                
                override fun onDataChannel(dc: DataChannel) {
                    dataChannel = dc
                    setupDataChannel()
                }
                
                override fun onIceConnectionChange(state: PeerConnection.IceConnectionState) {
                    Log.d(TAG, "ICE connection state: $state")
                }
                
                override fun onIceConnectionReceivingChange(receiving: Boolean) {
                    Log.d(TAG, "ICE connection receiving change: $receiving")
                }
                
                override fun onIceGatheringChange(state: PeerConnection.IceGatheringState) {
                    Log.d(TAG, "ICE gathering state: $state")
                }
                
                override fun onAddStream(stream: MediaStream) {
                    Log.d(TAG, "Remote stream added")
                }
                
                override fun onSignalingChange(state: PeerConnection.SignalingState) {}
                override fun onIceCandidatesRemoved(candidates: Array<out IceCandidate>?) {}
                override fun onRemoveStream(stream: MediaStream?) {}
                override fun onRenegotiationNeeded() {}
                override fun onAddTrack(receiver: RtpReceiver?, streams: Array<out MediaStream>?) {}
            }
        )
        
        // Create data channel for input events
        if (isSource) {
            val init = DataChannel.Init()
            dataChannel = peerConnection?.createDataChannel("input", init)
            setupDataChannel()
        }
    }
    
    /**
     * Setup data channel for input events
     */
    private fun setupDataChannel() {
        dataChannel?.registerObserver(object : DataChannel.Observer {
            override fun onBufferedAmountChange(amount: Long) {}
            
            override fun onStateChange() {
                Log.d(TAG, "Data channel state: ${dataChannel?.state()}")
            }
            
            override fun onMessage(buffer: DataChannel.Buffer) {
                if (isSource) {
                    // Receive input events from viewer
                    val data = ByteArray(buffer.data.remaining())
                    buffer.data.get(data)
                    val message = String(data)
                    handleInputEvent(message)
                }
            }
        })
    }
    
    /**
     * Create and send WebRTC offer
     */
    private fun createAndSendOffer() {
        val constraints = MediaConstraints().apply {
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveAudio", "false"))
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveVideo", "false"))
        }
        
        peerConnection?.createOffer(object : SdpObserver {
            override fun onCreateSuccess(sdp: SessionDescription) {
                peerConnection?.setLocalDescription(object : SdpObserver {
                    override fun onSetSuccess() {
                        sendOffer(sdp)
                    }
                    override fun onSetFailure(error: String) {
                        Log.e(TAG, "Failed to set local description: $error")
                    }
                    override fun onCreateSuccess(p0: SessionDescription?) {}
                    override fun onCreateFailure(p0: String?) {}
                }, sdp)
            }
            
            override fun onCreateFailure(error: String) {
                Log.e(TAG, "Failed to create offer: $error")
            }
            
            override fun onSetSuccess() {}
            override fun onSetFailure(error: String?) {}
        }, constraints)
    }
    
    /**
     * Send WebRTC offer via WebSocket
     */
    private fun sendOffer(sdp: SessionDescription) {
        Log.d(TAG, "=== SENDING WEBRTC OFFER ===")
        Log.d(TAG, "Session ID: $sessionId")
        Log.d(TAG, "Source Device ID (mobile): $sourceDeviceId")
        Log.d(TAG, "Viewer Device ID (laptop): $viewerDeviceId")
        Log.d(TAG, "SDP Type: ${sdp.type.canonicalForm()}")
        
        val message = JSONObject().apply {
            put("type", "webrtc_offer")
            put("sessionId", sessionId)
            put("deviceId", sourceDeviceId)
            put("payload", JSONObject().apply {
                put("toDevice", viewerDeviceId)
                put("data", JSONObject().apply {
                    put("type", sdp.type.canonicalForm())
                    put("sdp", sdp.description)
                })
                put("purpose", "remote_desktop")
            })
            put("timestamp", System.currentTimeMillis())
        }
        
        Log.d(TAG, "WebSocket ready state: ${if (webSocketManager != null) "connected" else "null"}")
        webSocketManager.sendMessage(message.toString())
        Log.d(TAG, "✅ WebRTC offer sent successfully")
    }
    
    /**
     * Send ICE candidate via WebSocket
     */
    private fun sendIceCandidate(candidate: IceCandidate) {
        val message = JSONObject().apply {
            put("type", "webrtc_ice_candidate")
            put("sessionId", sessionId)
            put("deviceId", if (isSource) sourceDeviceId else viewerDeviceId)
            put("payload", JSONObject().apply {
                put("toDevice", if (isSource) viewerDeviceId else sourceDeviceId)
                put("data", JSONObject().apply {
                    put("candidate", candidate.sdp)
                    put("sdpMid", candidate.sdpMid)
                    put("sdpMLineIndex", candidate.sdpMLineIndex)
                })
            })
            put("timestamp", System.currentTimeMillis())
        }
        
        webSocketManager.sendMessage(message.toString())
    }
    
    /**
     * Handle WebRTC signaling messages
     */
    fun handleSignaling(message: JSONObject) {
        val type = message.optString("type")
        val payload = message.optJSONObject("payload") ?: return
        val data = payload.optJSONObject("data") ?: return
        
        when (type) {
            "webrtc_offer" -> {
                // Viewer receives offer
                val sdp = SessionDescription(
                    SessionDescription.Type.OFFER,
                    data.optString("sdp")
                )
                peerConnection?.setRemoteDescription(object : SdpObserver {
                    override fun onSetSuccess() {
                        createAndSendAnswer()
                    }
                    override fun onSetFailure(error: String) {
                        Log.e(TAG, "Failed to set remote description: $error")
                    }
                    override fun onCreateSuccess(p0: SessionDescription?) {}
                    override fun onCreateFailure(p0: String?) {}
                }, sdp)
            }
            "webrtc_answer" -> {
                // Source receives answer
                val sdp = SessionDescription(
                    SessionDescription.Type.ANSWER,
                    data.optString("sdp")
                )
                peerConnection?.setRemoteDescription(object : SdpObserver {
                    override fun onSetSuccess() {
                        Log.d(TAG, "Remote description set successfully")
                    }
                    override fun onSetFailure(error: String) {
                        Log.e(TAG, "Failed to set remote description: $error")
                    }
                    override fun onCreateSuccess(p0: SessionDescription?) {}
                    override fun onCreateFailure(p0: String?) {}
                }, sdp)
            }
            "webrtc_ice_candidate" -> {
                val candidate = IceCandidate(
                    data.optString("sdpMid"),
                    data.optInt("sdpMLineIndex"),
                    data.optString("candidate")
                )
                peerConnection?.addIceCandidate(candidate)
            }
        }
    }
    
    /**
     * Create and send WebRTC answer
     */
    private fun createAndSendAnswer() {
        val constraints = MediaConstraints()
        peerConnection?.createAnswer(object : SdpObserver {
            override fun onCreateSuccess(sdp: SessionDescription) {
                peerConnection?.setLocalDescription(object : SdpObserver {
                    override fun onSetSuccess() {
                        sendAnswer(sdp)
                    }
                    override fun onSetFailure(error: String) {
                        Log.e(TAG, "Failed to set local description: $error")
                    }
                    override fun onCreateSuccess(p0: SessionDescription?) {}
                    override fun onCreateFailure(p0: String?) {}
                }, sdp)
            }
            override fun onCreateFailure(error: String) {
                Log.e(TAG, "Failed to create answer: $error")
            }
            override fun onSetSuccess() {}
            override fun onSetFailure(error: String?) {}
        }, constraints)
    }
    
    /**
     * Send WebRTC answer via WebSocket
     */
    private fun sendAnswer(sdp: SessionDescription) {
        val message = JSONObject().apply {
            put("type", "webrtc_answer")
            put("sessionId", sessionId)
            put("deviceId", viewerDeviceId)
            put("payload", JSONObject().apply {
                put("toDevice", sourceDeviceId)
                put("data", JSONObject().apply {
                    put("type", sdp.type.canonicalForm())
                    put("sdp", sdp.description)
                })
            })
            put("timestamp", System.currentTimeMillis())
        }
        
        webSocketManager.sendMessage(message.toString())
        Log.d(TAG, "Sent WebRTC answer")
    }
    
    /**
     * Handle input event from viewer
     */
    private fun handleInputEvent(message: String) {
        try {
            val json = JSONObject(message)
            val eventType = json.optString("type")
            Log.d(TAG, "Received input event: $eventType")
            
            // Note: Android doesn't allow programmatic input injection without root
            // This would require accessibility service or root access
            // For now, just log the events
        } catch (e: Exception) {
            Log.e(TAG, "Failed to handle input event", e)
        }
    }
    
    /**
     * Stop screen sharing
     */
    fun stopScreenShare() {
        try {
            Log.d(TAG, "Stopping screen share...")
            
            // Stop video capturer first
            videoCapturer?.let {
                try {
                    it.stopCapture()
                    Log.d(TAG, "Video capturer stopped")
                } catch (e: Exception) {
                    Log.e(TAG, "Error stopping video capturer", e)
                }
            }
            
            // Close data channel
            dataChannel?.let {
                try {
                    it.close()
                    Log.d(TAG, "Data channel closed")
                } catch (e: Exception) {
                    Log.e(TAG, "Error closing data channel", e)
                }
            }
            dataChannel = null
            
            // Close peer connection
            peerConnection?.let {
                try {
                    it.close()
                    Log.d(TAG, "Peer connection closed")
                } catch (e: Exception) {
                    Log.e(TAG, "Error closing peer connection", e)
                }
            }
            peerConnection = null
            
            // Dispose video track
            videoTrack?.let {
                try {
                    it.setEnabled(false)
                    it.dispose()
                    Log.d(TAG, "Video track disposed")
                } catch (e: Exception) {
                    Log.e(TAG, "Error disposing video track", e)
                }
            }
            videoTrack = null
            
            // Dispose video source
            videoSource?.let {
                try {
                    it.dispose()
                    Log.d(TAG, "Video source disposed")
                } catch (e: Exception) {
                    Log.e(TAG, "Error disposing video source", e)
                }
            }
            videoSource = null
            
            // Dispose video capturer
            videoCapturer?.let {
                try {
                    it.dispose()
                    Log.d(TAG, "Video capturer disposed")
                } catch (e: Exception) {
                    Log.e(TAG, "Error disposing video capturer", e)
                }
            }
            videoCapturer = null
            
            // Dispose local stream
            localStream?.let {
                try {
                    it.dispose()
                    Log.d(TAG, "Local stream disposed")
                } catch (e: Exception) {
                    Log.e(TAG, "Error disposing local stream", e)
                }
            }
            localStream = null
            
            // Note: MediaProjection is NOT stored as a class variable anymore
            // It's created fresh each time in startScreenShare() to avoid reuse errors
            // The MediaProjection will be stopped automatically when the ScreenCapturerAndroid is disposed
            
            Log.d(TAG, "Screen sharing stopped successfully")
        } catch (e: Exception) {
            Log.e(TAG, "Error in stopScreenShare", e)
        }
    }
    
    /**
     * Cleanup resources
     */
    fun cleanup() {
        stopScreenShare()
    }
}
